src/Hazelcast.Net/CP/CpSubSystemSession.Heartbeat.cs

```diff
@@ -0,0 +1,102 @@
+﻿// Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Hazelcast.Core;
+using Microsoft.Extensions.Logging;
+
+namespace Hazelcast.CP
+{
+    /// <summary>
+    /// Implements Heartbeat side of CP Session
+    /// </summary>
+    internal partial class CpSubsystemSession
+    {
+        private readonly CancellationTokenSource _cancel;
+        private Task _heartbeating;
+        private int _heartbeatState;
+
+        /// <summary>
+        /// Schedules the session heatbet requests. It does not require to check state of the heartbeat
+        /// <para>Use only the method to run heartbeat.</para>
+        /// </summary>
+        protected void ScheduleHeartbeat(TimeSpan period)
+        {
+            if (_disposed == 0 && Interlocked.CompareExchange(ref _heartbeatState, 1, 0) == 0)
+            {
+                _heartbeating = BeatAsync(period, _cancel.Token);
+            }
+        }
+
+        private async Task BeatAsync(TimeSpan period, CancellationToken cancellationToken)
+        {
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                await Task.Delay(period, cancellationToken).CfAwait();
+
+                if (cancellationToken.IsCancellationRequested) break;
+
+                await RunAllAsync(cancellationToken).CfAwait();
+            }
+        }
+
+        /// <summary>
+        /// Runs for all sessions
+        /// </summary>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        private async Task RunAllAsync(CancellationToken cancellationToken)
+        {
+            _logger.LogDebug("Run CP Session Heartbeat");
+
+            lock (_mutex) if (_sessions.IsEmpty) return;
+
+            IEnumerable<(CPGroupId, CPSubsystemSessionState)> sessions;
+            lock (_mutex) sessions = _sessions.Select(p => (p.Key, p.Value));
+
+            await sessions.ParallelForEachAsync((p, cancellationToken) =>
```

Comment:
.NET 6 introduces [Parallel.ForEachAsync](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.parallel.foreachasync?view=net-6.0) and we should follow this syntax to be future-proof. But we cannot add a static method to `Parallel` for pre- .NET 6. In other situations what we've done is create a `ParallelEx` class in the `Polyfills` namespace, and just copy/paste the .NET 6 code there (backporting, in other words). You'll find the code [here](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Linq.Parallel/src/System/Linq/ParallelEnumerable.cs). The benefit is that... we trust that code and don't write tests for it, since it's MS code and supposed to work.

So here it would be... `ParallelEx.ForEachAsync(...)` using the syntax that would be valid in .NET 6.

Reply:
I found the `Parallel.ForeachAsync` at [here](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/Parallel.ForEachAsync.cs) However, it requires nested things to be ported as well. So, I'm canceling the custom parallel for each for now since it needs more time than we think.